File: backend/src/models/query.py
from __future__ import annotations

from typing import Any

from pydantic import Field

from .base import AppBaseModel


class QueryRequest(AppBaseModel):
    sql_text: str = Field(..., min_length=1)
    timeout_seconds: int = Field(30, ge=1)
    max_rows: int = Field(1000, ge=1)


class QueryColumn(AppBaseModel):
    name: str
    type: str


class QueryResponse(AppBaseModel):
    columns: list[QueryColumn]
    rows: list[list[Any]]
    duration_ms: int
    limit_applied: int
    request_id: str

File: frontend/src/services/query.ts
import { apiFetch } from "./api";

export interface QueryRequest {
  sqlText: string;
  timeoutSeconds?: number;
  maxRows?: number;
}

export interface QueryColumn {
  name: string;
  type: string;
}

export interface QueryResponse {
  columns: QueryColumn[];
  rows: unknown[][];
  durationMs: number;
  limitApplied: number;
  requestId: string;
}

export function executeQuery(connectionId: string, payload: QueryRequest): Promise<QueryResponse> {
  return apiFetch(`/connections/${connectionId}/query`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
